packages/truffle-db/src/loaders/artifacts/index.ts

```diff
@@ -287,21 +381,98 @@ export class ArtifactsLoader {
+              const transaction = await web3.eth.getTransaction(filteredNetwork[0][1]["transactionHash"]);
+              const historicBlock = {
+                height: transaction.blockNumber,
+                hash: transaction.blockHash
+              }
+
+              const networksAdd = await this.db.query(AddNetworks,
+              {
+                networks:
+                [{
+                  name: network,
+                  networkID: networkID,
+                  historicBlock: historicBlock
+                }]
+              });
+
+              const networkId = networksAdd.data.workspace.networksAdd.networks[0].id;
+              configNetworks.push({
+                contract: contractName,
+                id: networkId,
+                address: filteredNetwork[0][1]["address"]
+              });
+            }
+          }
+        }
+      }
+      return configNetworks;
+    }));
+    return networksByContract;
+  }
+
+  async loadContractInstances (contracts: Array<ContractObject>, contractIds: Array<IdObject>, networksArray: Array<Array<LoaderNetworkObject>>) {
+    // networksArray is an array of arrays of networks for each contract;
+    // this first mapping maps to each contract
+    const instances = networksArray.map((networks, index) => {
+      // this second mapping maps each network in a contract
+      const contractInstancesByNetwork = networks.map((network) => {
+        let instance = {
+          address: network.address,
+          contract: contractIds[index],
+          network: {
+            id: network.id
+          }
+        }
+        return instance;
+      });
+
+      return contractInstancesByNetwork;
+    });
+
+    await this.db.query(AddContractInstances, { contractInstances: instances.flat() });
+  }
+
+  async loadCompilation (compilationConfig: CompilationConfigObject) {
     const compilationOutput = await Contracts.compile(compilationConfig);
     const contracts = await this.organizeContractsByCompiler(compilationOutput);
-
     const compilationObjects = await Promise.all(Object.values(contracts)
       .filter(contractsArray => contractsArray.length > 0)
-      .map(contractsArray =>
-        this.setCompilation(contractsArray))
-    );
+      .map(contractsArray => this.setCompilation(contractsArray)));
 
 
     const compilations = await this.db.query(AddCompilation, { compilations: compilationObjects });
 
-    //map contracts to the compilation they belong in before adding them to truffle-db
-    await Promise.all(compilations.data.workspace.compilationsAdd.compilations
-      .map(({ compiler, id }) => this.loadCompilationContracts(contracts[compiler.name], id)));
+    //map contracts and contract instances to compiler
+    await Promise.all(compilations.data.workspace.compilationsAdd.compilations.map(async ({compiler, id}) => {
+      const contractIds = await this.loadCompilationContracts(contracts[compiler.name], id, compiler.name);
```

Comment:
@gnidan I remember you mentioned that you thought I should pull this part out into its own function, but it's using from this loadCompilations function so I wanted to double check what you mean here. Thanks!

Reply:
I made some changes here in line with this pseudo code. I can break it down some more but I don't want to have to map an extra time if it can be avoided. Let me know if this is more in line with what you imagined or if you'd like additional changes!